Stack reconstruction after guard failure.

ptersilie opened the pull request:
It's finally here. The long awaited stack reconstruction PR. It's not perfect, there's likely still bugs, and I'm sure some of the comments could be improved. But I've been dragging this branch around for way to long, so it's time to get it in and iterate on top of that. Naturally, this passes all the tests, and we can even run a simple Lua program.

There's a companion PR coming too which adds prologue information (e.g. callee saved registers) to stackmaps and is needed by this PR. But I haven't finished cleaning that up yet, so we can start working through the ~800 changes in this PR while we wait. I'm sure there'll be loads to discuss and unpack here. Enjoy!

Companion PR: https://github.com/ykjit/ykllvm/pull/48

ltratt commented:
Is there any use in reviewing the 4 commits separately?

ptersilie commented:
Probably not. Want me to squash first?

ptersilie commented:
Squashed.

vext01 commented on tests/c/debug_debuginfo.c:
```diff
@@ -3,7 +3,7 @@
 //   stderr:
 //     ...
-//     define i8 @__yk_compiled_trace_0(...
+//     define ptr @__yk_compiled_trace_0(...
```
Why did the trace return type change again?

vext01 replied:
A later change explains, I think:

>  Returns the address of a reconstructed stack or null if there wasn't a guard failure.

vext01 commented on tests/c/interp_inside_call2.c:
```diff
@@ -12,6 +12,7 @@
 //     jit-state: enter-stopgap
 //     ...
 //     jit-state: exit-stopgap
+//     jit-state: exit-jit-code
```
Should the `jit-state: enter-stopgap` and `jit-state: exit-stopgap` states be replaced by a single `jit-state: deoptimise`?

ptersilie replied:
Not sure what advantage this would have. We know we are deoptimising when we see `enter-stopgap/exit-stopgap`.

vext01 replied:
But we don't stopgap, do we?

ptersilie replied:
I misunderstood you. You want to rename it. I guess we can. Is what we are doing now not stopgapping anymore? Just looking at the definition of stopgap I would still say it fits.

vext01 replied:
Regardless of what you call it, whereas before stopgapping was an interval of time (in which we happened to be running an interpreter), now it's an instantaneous event. We flick from JITted code back to the interpreter without any intermediate "user program execution". So to start/stop stopgapping makes little sense in my head.

And I do still think it should be called "deoptimise" and not "stopgap". The name "stopgap" implies a temporary compromise, which no longer conveys what we are really doing any more.

ltratt replied:
Edd's summed up my position well.

ptersilie replied:
I'm not attached to stopgapping and get your point. So I'm happy to rename it.

ltratt replied:
Correct: "guard failure" / "deoptimisation" (they're the same thing for us) is effectively a discreet event.

ptersilie replied:
Well this wasn't fun. [303a233](https://github.com/ykjit/yk/pull/578/commits/303a23327920839ec2c4f03f89b9c7c455029e3d)

ltratt replied:
But now it's done :)

vext01 commented on ykcapi/src/lib.rs:
```diff
@@ -41,22 +41,22 @@ pub extern "C" fn yk_mt_control_point(_mt: *mut MT, _loc: *mut Location) {
 }
 
 // The "real" control point, that is called once the interpreter has been patched by ykllvm.
+// Returns the address of a reconstructed stack or null if there wasn#t a guard failure.
```
wasn#t

vext01 replied:
How does the reconstructed stack get used? It seems that a pointer to it is returned directly to the caller. What does the caller then do with it?

ptersilie replied:
This is also explain better in the commit message. The caller checks the value of the return and if it's not null, then it jumps to `__ykrt_reconstruct_frames` which copies over the new stack.

vext01 commented on ykcapi/src/lib.rs:
```diff
@@ -147,20 +147,70 @@ pub struct LiveAOTVals {
+            // over. Move it into rdx.
+            "mov rdx, [rdi]",
+            // Then adjust the address to where the new stack actually starts.
+            "add rdi, 8",
```
Not sure I understand this bit. rdi containst the size of the map, which we then add 8 to, but the second instruction is talking about an address. What am i missing?

ptersilie replied:
We first copy the size of the map into `rdx`: `mov rdx, [rdi]`.
Then we add 8 to move the pointer of `rdi` to the actual beginning of the new stack. We then use that adjusted `rdi` as an argument for `memcpy` in order to copy over the new stack frames.

vext01 replied:
ah, i see what's gone wrong. I thought you were adding to rdx. My bad.

vext01 commented on ykcapi/src/lib.rs:
```diff
@@ -147,20 +147,70 @@ pub struct LiveAOTVals {
+            "pop rbx",
+            "pop rcx",
+            "pop rdx",
+            "pop rax",
```
Can we put a FIXME in here to remind us to restore floating point registers at some stage?

ptersilie replied:
Fixed in [7660a3c](https://github.com/ykjit/yk/pull/578/commits/7660a3c9813795998bfb6a958058c2d30b60fc9f)

vext01 commented on ykcapi/src/lib.rs:
```diff
@@ -178,9 +228,9 @@ pub extern "C" fn yk_stopgap(
     let registers = Registers::from_ptr(rsp);
 
-    let mut sginterp = unsafe { SGInterp::new(activeframes, retvalptr) };
+    let mut sginterp = unsafe { SGInterp::new(activeframes, frameaddr) };
```
For some reason I was expecting yksgi to disappear. Are we keeping it?

ptersilie replied:
All the initialisation code from SGI is still needed for stack reconstruction as well. I'm also a bit hesitant to remove SGI entirely as it might be useful in the future for debugging purposes. I imagined that at some point we have a switch to toggle stack reconstruction/stopgap interpreter.

vext01 replied:
I find it confusing that we instantiate a stop gap interpreter, but then don't *really* use it.

ltratt replied:
I tend to agree with Edd. If we do ever need a stopgap interpreter, we'll probably have to rewrite more-or-less from scratch anyway. I'd be inclined to remove it entirely.

vext01 replied:
There's always the git history.

ltratt replied:
Agreed. I'm going to resolve this particular comment but let's assume that we rip out most of the SGI in an upcoming PR.

vext01 commented on ykcapi/src/lib.rs:
```diff
@@ -218,6 +268,7 @@ pub extern "C" fn yk_stopgap(
             SMLocation::Indirect(reg, off, size) => {
                 let addr = unsafe { registers.get(*reg) as *mut u8 };
                 let addr = unsafe { addr.offset(isize::try_from(*off).unwrap()) };
+                //println!("direct addr: {:?}", addr);
```
debugging left in here.

ptersilie replied:
Fixed in [c45e5e6](https://github.com/ykjit/yk/pull/578/commits/c45e5e644b2bc6b43c9d456a8192a5cc3ddbaff8)

vext01 commented on ykllvmwrap/src/jitmodbuilder.cc:
```diff
@@ -1448,8 +1441,9 @@ class JITModBuilder {
 
-    Builder.CreateRet(ConstantInt::get(Type::getInt8Ty(JITMod->getContext()),
-                                       TRACE_RETURN_SUCCESS));
+    // If the trace succeeded return a nullpointer instead of a reconstructed
```
nitpicking: `nullpointer` -> `null pointer`. Sorry :P

ptersilie replied:
Fixed in [c45e5e6](https://github.com/ykjit/yk/pull/578/commits/c45e5e644b2bc6b43c9d456a8192a5cc3ddbaff8)

vext01 commented on yksgi/Cargo.toml:
```diff
@@ -9,7 +9,15 @@ license = "Apache-2.0 OR MIT"
+memmap2 = "0.5.2"
+capstone = "0.11.0"
+
+[dependencies.object]
```
What is `object` needed for?

vext01 replied:
I answered this one myself in a later comment.

vext01 commented on yksgi/src/lib.rs:
```diff
@@ -8,19 +8,25 @@ use std::collections::HashMap;
 use std::convert::TryFrom;
 use std::ffi::{c_void, CStr};
 use std::ptr;
+
```
Any comments on this line @ltratt? :)

ltratt replied:
I think... let me check... yes, this is extraneous space isn't it?

ltratt replied:
Being serious, we can collapse these into `use std::{convert.., ffi::.., }` and so on rather than having Java-esque multi-line `use`s.

ptersilie replied:
I haven't formatted yet. So hopefully the formatter will pick that up. But will double check it does.

ltratt replied:
The formatter probably won't merge them, but I might be wrong. [Let's also run the formatter before raising the PR :)]

ptersilie replied:
Yeah, I may have rushed this PR out a bit to quickly, but it was late last night and I wanted to wrap it up.

ltratt replied:
Let's manually merge the `use`s now and be done with it?

ptersilie replied:
Fixed in 3724ca6647067bc052cb351907bdff5dd114d2e1.

vext01 commented on yksgi/src/lib.rs:
```diff
@@ -149,6 +168,271 @@ impl SGInterp {
 
+    // Generate frames from stackmap information after a guard failure. The new frames are stored
+    // inside some allocated memory whose pointer this function returns. The frames are then later
+    // `memcpy`ed to the actual stack by ykcapi::__ykrt_reconstruct_frames.
```
another nitpick: backtick `ykcapi::__ykrt_reconstruct_frames`

ltratt replied:
These should also be `///` doc string quotes. And then `[ykcapi::__ykrt_reconstruct_frames]` (i.e. put it in backticks) and rustdoc will  make it a clickable link! [I don't know, but it wouldn't surprise me if this also allows editors to make this a followable link.]

ptersilie replied:
Will fix.

ptersilie replied:
Fixed in [c45e5e6](https://github.com/ykjit/yk/pull/578/commits/c45e5e644b2bc6b43c9d456a8192a5cc3ddbaff8)

vext01 commented on yksgi/src/lib.rs:
```diff
@@ -149,6 +168,271 @@ impl SGInterp {
+        // Load the stackmap and text sections from the binary. The former we need to parse in the
+        // stackmaps. The latter is required to disassemble the binary in order to distinguish
+        // branches from calls.
+        // FIXME: Don't use current_exe.
```
OK, this is where object is used. We shouldn't need to look into the binary file for either `.text` or the stackmap section. They are both loaded now. This could be a separate PR if @ltratt is OK with that.

We might also be able to avoid instantiating capstone using some magic. What do you need? the size of the instruction in the AOT code which could become a guard failure in the JIT code?

ptersilie replied:
Yes I probably need @vext01 help to get rid of the `current_exe`. Will do a follow up PR immediately once this is in.

Regarding the capstone thing. I need to size of the instruction before stackmap. Which may be a `call` or an `icmp` or `test`. If it's a call, I need its size in order to adjust the record offset.

vext01 replied:
We may be able to insert something at compile-time, but let's roll with this for now.

Perhaps add a FIXME?

vext01 replied:
I'd say roll with capstone and deal with a better way in a follow up PR.

ltratt replied:
I seem to be saying this a lot, but I agree with Edd.

ptersilie replied:
I concur. This will likely require a ykllvm change too. So better for another time.

ptersilie replied:
Added a fixme: [9294cdb](https://github.com/ykjit/yk/pull/578/commits/9294cdb1fea1440e857e172f981b9a5886501ac2)

vext01 commented on yksgi/src/lib.rs:
```diff
@@ -149,6 +168,271 @@ impl SGInterp {
+        // so we can figure out the size of the instruction we
+        // return too in order to adjust the return address in our reconstructed stackmap.
+        let cs = Capstone::new()
+        .x86()
```
weird indent.

ptersilie replied:
Fixed in 0c4cc76c6098d90ce398f5ea2e4463c5b8536a8b

vext01 commented on yksgi/src/lib.rs:
```diff
@@ -149,6 +168,271 @@ impl SGInterp {
+
+        // Create a capstone dissassembler.
+        // so we can figure out the size of the instruction we
+        // return too in order to adjust the return address in our reconstructed stackmap.
```
too -> to

ptersilie replied:
Fixed in [c45e5e6](https://github.com/ykjit/yk/pull/578/commits/c45e5e644b2bc6b43c9d456a8192a5cc3ddbaff8)

vext01 commented on yksgi/src/lib.rs:
```diff
@@ -149,6 +168,271 @@ impl SGInterp {
+        for (i, frame, rec, pinfo) in smaps {
+            debug_assert!(rec.size != u64::MAX);
+
+            // WRITE RBP
```
WRITE -> Write

ltratt commented on ykcapi/src/lib.rs:
```diff
@@ -278,7 +329,8 @@ pub extern "C" fn __llvm_deoptimize(
     unsafe {
         asm!(
-            // Save registers to the stack.
+            // Save registers the may be referenced by the stackmap to the stack before they get
```
Sentence doesn't parse. I'm not sure if it's supposed to be "THAT may be referenced..."?

ptersilie replied:
Sorry, yes. It's supposed to be `Save registers that`. Will fix.

ptersilie replied:
Fixed in [c45e5e6](https://github.com/ykjit/yk/pull/578/commits/c45e5e644b2bc6b43c9d456a8192a5cc3ddbaff8)

ltratt commented on ykllvmwrap/src/jitmodbuilder.cc:
```diff
@@ -1312,6 +1303,9 @@ class JITModBuilder {
                 (IID == Intrinsic::lifetime_end))
               continue;
 
+            if (IID == Intrinsic::frameaddress)
```
Why do we skip this one?

ptersilie replied:
This isn't needed in the compiled trace as it's only passed into the control point which goes away as well in a compiled trace.

ltratt replied:
Can we add a quick comment for our future selves?

ptersilie replied:
Fixed in [f689da1](https://github.com/ykjit/yk/pull/578/commits/f689da1d385791e1c7a821174b1712ddb0dd18b9)

ltratt commented on yksgi/src/lib.rs:
```diff
@@ -98,6 +104,19 @@ signext!(sext_i8, i8);
 signext!(sext_i16, i16);
 signext!(sext_i32, i32);
 
+unsafe fn get_stackmap_call(pc: Value) -> Value {
```
Why is the whole function `unsafe`?

ptersilie replied:
Tbh I'm still slightly confused about `unsafe` and where to use it. I think my reasoning was that depending on which value of pc you pass in this might not work. So it's on the caller to make sure the argument is safe.

ltratt replied:
In some sense *any* function won't work if you call it incorrectly! In our context `unsafe` should mostly mean "if you call this with the wrong arguments the program may exhibit undefined behaviour (inc. but not limited to segfaulting)". Is this one of those functions?

ptersilie replied:
Not sure about segfaulting. I'm inheriting the unsafe from the llvm-sys calls (e.g. `LLVMGetPreviousInstruction`). Not entirely certain how you can get them to segfault but I imagine handing them a nullptr would do the trick.

So as long as we don't do that we should mark the LLVM-calls unsafe instead of the whole function?

ptersilie replied:
The `get_stackmap_call` isn't checking any preconditions. I don't know what might result in a segfault or UB in `LLVMGetPreviousInstruction`, so I wouldn't know what to check for. So maybe marking the function as `unsafe` is fine here?

ptersilie replied:
Okay, I had a poke around in the LLVM API sources to see where there is potential for UB. It looks like as long as you pass the API the correct type its expecting things should be fine. So I've added some extra asserts to make sure we are not passing nonsense and removed the unsafe from the function. Does that sound ok? See [5993877](https://github.com/ykjit/yk/pull/578/commits/599387756b3ae04da9bccab4d02c586b72e52a7a)

ltratt replied:
Perfect!

vext01 commented:
Good work @ptersilie. I'm willing to bet this is the scariest low-level code you've ever written to date!

ltratt commented on yksgi/src/lib.rs:
```diff
@@ -149,6 +168,271 @@ impl SGInterp {
+
+        // The final size of the memory we need to allocate. Inititialised with space to store
+        // registers for register recovery.
+        let mut memsize: usize = 15 * 8;
```
8 -> `sizeof::<*const ()>` or similar.

ptersilie replied:
Fixed in [3e270bc](https://github.com/ykjit/yk/pull/578/commits/3e270bceda6db7073f58c4fd90cea45ea6bcadab)

ltratt commented on yksgi/src/lib.rs:
```diff
@@ -149,6 +168,271 @@ impl SGInterp {
+    // Generate frames from stackmap information after a guard failure. The new frames are stored
+    // inside some allocated memory whose pointer this function returns. The frames are then later
+    // `memcpy`ed to the actual stack by ykcapi::__ykrt_reconstruct_frames.
+    pub fn reconstruct_frames(&self, btmframeaddr: *mut c_void) -> *const c_void {
```
This function currently seems very x86 only. Should it be guarded as such?

ltratt commented on yksgi/src/lib.rs:
```diff
@@ -149,6 +168,271 @@ impl SGInterp {
+            let rec = rec.unwrap();
+            let pinfo = pinfo.unwrap();
+            if i > 0 {
+                // We don't need to include the bottom-most frame, which is the frame containing
```
I don't understand how the comment relates to the `memsize +=`.

ptersilie replied:
Would it help if it was above the condition? The comment explains why we skip `i=0`.

ptersilie replied:
Fixed in [c45e5e6](https://github.com/ykjit/yk/pull/578/commits/c45e5e644b2bc6b43c9d456a8192a5cc3ddbaff8)

ptersilie replied:
Got you. So `rec.size` is the size of each stack frame. In order to calculate how much memory we need to allocate for out new stack, we add the frame sizes of all frames together. However, the stackmap (which we iterate over) includes the bottom frame (the caller of the control point). But since that frame already exists, we don't need to allocate space for it. Does that clear things up?

ptersilie replied:
See if [1f2afe8](https://github.com/ykjit/yk/pull/578/commits/1f2afe8397349af56993f7de0c1e1152aa212eac) works better for you.

ltratt replied:
Got it!

ltratt commented on yksgi/src/lib.rs:
```diff
@@ -149,6 +168,271 @@ impl SGInterp {
+
+            // Update RBP to represent this frame's address.
+            if pinfo.hasfp {
+                registers[6] = currframe as u64;
```
Is this cast always safe?

ptersilie replied:
On x86_64 yes. Let me see if I can turn them into usizes.

ptersilie replied:
I could probably make `registers` store `usize`, but we are storing `SGValue.val` in them, which is `u64`. We had a discussion some time ago that `SGValue` probably shouldn't store `u64`s. But I don't think that's a change for this PR, and I'm also unsure how high on the list this should go.

ltratt replied:
I was thinking the other way around. I understand that on x86_64 the registers can be treated as `u64`: I just wasn't entirely sure if `currframe` could be case to a `u64` in all cases or not. [I'm not sure what type `currframe` has. Knowing that might make it obvious.]

ptersilie replied:
Does this explanation resolve this for you?

ltratt commented on yksgi/src/lib.rs:
```diff
@@ -149,6 +168,271 @@ impl SGInterp {
+                    val.val
+                } else {
+                    todo!()
+                    //println!("probably return pointer which will disappear");
```
Commented code (you can put text in `todo` if you want e.g. `todo!("probably return a pointer which will disappear")`).

ptersilie replied:
Actually, I need to reinvestigate this code here. I think this can be removed entirely now.

ptersilie replied:
[bc9be2b](https://github.com/ykjit/yk/pull/578/commits/bc9be2bb1a42334c625104ce5309e9a7b5f0cb8b)

ltratt commented on yksgi/src/lib.rs:
```diff
@@ -149,6 +168,271 @@ impl SGInterp {
+                        }
+                        debug_assert!(op.is_alloca());
+
+                        // The sizes reported by the stackmap aren't always correct. But we can get
```
Shouldn't we fix stackmaps then? [Not in this PR, but in general.]

vext01 replied:
I'm not sure it's worth the effort right now to be honest.

ltratt replied:
Maybe we should raise an issue? At the moment it looks like we're doing work at run-time that we *could* semi-easily shift to AOT.

vext01 replied:
https://github.com/ykjit/yk/issues/579

ltratt replied:
Thanks!

ltratt commented on yksgi/src/lib.rs:
```diff
@@ -149,6 +168,271 @@ impl SGInterp {
+                        // The sizes reported by the stackmap aren't always correct. But we can get
+                        // the correct size from the IR.
+                        let eltype = unsafe { LLVMGetAllocatedType(op.get()) };
+                        let size = i64::try_from(unsafe { LLVMABISizeOfType(layout, eltype)}).unwrap();
```
Do we need a signed size here?

ptersilie replied:
Can allocas have negative sizes?

ptersilie replied:
Oh, I see. You are saying we don't need an `i64` here. And I agree.

ltratt replied:
What type does `LLVMABISizeOfType` return? I wonder if the `try_from` is even needed now.

ptersilie replied:
Good spot. It's `c_ulonglong` which means we don't need the `try_from`.

Fixed in [cfe5de7](https://github.com/ykjit/yk/pull/578/commits/cfe5de756bc0bd05c3f44dc7a0298e3a5ad1233a)

ltratt commented on yksgi/src/lib.rs:
```diff
@@ -149,6 +168,271 @@ impl SGInterp {
+                    rsp = unsafe { rsp.offset(8)};
+                }
+            }
+            // WRITE RETURN ADDRESS.
```
Not sure we need this shouty line :)

ptersilie replied:
I did the caps on purpose to divide this code into sections and make them easier to find. Each caps title writes a different section to the new stack.

ltratt commented on yksgi/src/lib.rs:
```diff
@@ -547,6 +831,7 @@ impl SGInterp {
             None
         } else {
             let op = self.pc.get_operand(0);
+            println!("ret op: {:?}", op.as_str());
```
Dugging print.

ltratt commented on yksmp/src/lib.rs:
```diff
@@ -14,11 +14,15 @@ use std::error;
 struct Function {
     addr: u64,
     record_count: u64,
+    stack_size: u64,
```
Can we not make this `usize` and de-x86 it?

ptersilie replied:
Just checked, this is already u64 in LLVM, so we inherit the type here. https://llvm.org/docs/StackMaps.html#stackmap-format

ltratt replied:
OK.

ltratt commented on yksmp/src/lib.rs:
```diff
@@ -14,11 +14,15 @@ use std::error;
-struct Record {
-    offset: u32,
-    live_vars: Vec<LiveVar>,
+pub struct Record {
```
Similarly these are all very x86 centric, probably unnecessarily so?

ptersilie replied:
`size` we later populate with the functions stack size which is also u64 in LLVM's stackmap format.

ptersilie commented on yksgi/src/lib.rs:
```diff
@@ -149,6 +168,271 @@ impl SGInterp {
+        // Get the modules layout which we'll need to extract type sizes of LLVM IR.
+        let layout = self.module.datalayout();
+
+        // Generate and write frames to the mmap. Since the stack grows downwards and we need to
```
If it isn't already I will guard it.

ptersilie replied:
Fixed in [fb53ebe](https://github.com/ykjit/yk/pull/578/commits/fb53ebedc5a38dfb9dc4e330a6a906fdf4f4c073)

ptersilie commented on yksgi/src/lib.rs:
```diff
@@ -149,6 +168,271 @@ impl SGInterp {
+                }
+            }
+
+            // WRITE STACKMAP LOCATIONS.
```
I did the caps on purpose to divide this code into sections and make them easier to find. You don't seem to be a fan.

vext01 replied:
I don't mind as long as it is consistent.

ltratt replied:
I don't like the shouting either :)

ptersilie commented on yksgi/src/lib.rs:
```diff
@@ -149,6 +168,271 @@ impl SGInterp {
+                    val.val
+                } else {
+                    todo!()
+                    //println!("probably return pointer which will disappear");
```
Fixed in [c45e5e6](https://github.com/ykjit/yk/pull/578/commits/c45e5e644b2bc6b43c9d456a8192a5cc3ddbaff8)

ptersilie commented:
Think this is ready for re-review.

vext01 commented:
The main outstanding thing with this is the confusion around a "stop gap interpreter"

Is it worth pulling the frame initialisation code out of yksgi?

ltratt commented on yksgi/src/lib.rs:
```diff
@@ -30,6 +30,9 @@ static ALWAYS_SKIP_FUNCS: [&str; 8] = [
 static SKIP_FOR_NOW_FUNCS: [&str; 2] = ["yk_location_drop", "yk_mt_drop"];
 static YK_CONTROL_POINT_FUNC: &str = "__ykrt_control_point";
 
+static POINTER_USIZE: usize = std::mem::size_of::<*const ()>();
```
Slightly confusing variable name. `SIZEOF_POINTER` would be better. [Personally I'd just `use std::mem::size_of` and then always use `size_of::<*const ()>` in the code. I don't care too much though.]

ptersilie replied:
I would still need to have a constant as an `isize` tho unless I want to do `isize::try_from(size_of::<*const ()>).unwrap()` everywhere, which IMO distracts from the code quite a bit.

ptersilie replied:
Can I do two constants: `ISIZEOF_POINTER` and `USIZEOF_POINTER`?

ltratt replied:
I prefer size_of, but it's something upon which reasonable people can differ. So long as we change the variable name to something less confusing, I'm OK with a variable.

ltratt replied:
> Can I do two constants: ISIZEOF_POINTER and USIZEOF_POINTER?

If you need them, of course!

ptersilie replied:
I think I'd prefer two as having them as constants makes the main code much more readable. Renamed them in [5993877](https://github.com/ykjit/yk/pull/578/commits/599387756b3ae04da9bccab4d02c586b72e52a7a)

ltratt replied:
Wrong commit?

ltratt replied:
Aha it should be this one https://github.com/ykjit/yk/pull/578/commits/90ce72e5904bc878733751e61e90ac82f839d302

ptersilie replied:
Oh sorry.

ltratt commented:
I've done a round of commenting / resolving, including comments that @vext01 raised that I was trivially able to confirm are resolved.

ptersilie commented:
> The main outstanding thing with this is the confusion around a "stop gap interpreter"
> 
> Is it worth pulling the frame initialisation code out of yksgi?

Not sure this is easily done. All the initialisation stuff on yksgi (e.g. `var_init`) is needed for stack reconstruction. So unless `yksgi` provides an API so lookup variables I don't think this is easy to pull out.

ltratt commented:
@ptersilie I think we should keep this PR as-is, but maybe an upcoming PR can rip out the bits of the stopgap interpreter we're never likely to need again.

ltratt commented:
I've done another round of resolving -- we're getting there!

ptersilie commented:
Think this might be ready for a (hopefully final) re-review.

ltratt commented on yksgi/src/lib.rs:
```diff
@@ -278,7 +278,7 @@ impl SGInterp {
             // the bottom-most frame).
             if pinfo.hasfp && i > 0 {
-                rsp = unsafe { rsp.offset(-POINTER_ISIZE) };
+                rsp = unsafe { rsp.offset(-ISIZEOF_POINTER) };
```
Isn't this `rsp.sub(USIZEOF_POINTER)` in disguise? [I tend to prefer avoiding negative `isize`s because we inevitably end up having usize <-> isize conversions and those are surprisingly easy to get wrong.]

ptersilie replied:
I didn't know there was a `sub` call. :man_facepalming: Fixed in cf3e4be3b4da5c85522c35649d24fd7e155dd76a.

Sadly this (in combination with add) doesn't allow us to get rid of the ISIZE_POINTER constant, since we still have some places where we get passed an isize offset.

ltratt replied:
OK.

ptersilie commented:
Also found a bug which I introduced in [cf3e4be](https://github.com/ykjit/yk/pull/578/commits/cf3e4be3b4da5c85522c35649d24fd7e155dd76a). Fixed in [0c4cc76](https://github.com/ykjit/yk/pull/578/commits/0c4cc76c6098d90ce398f5ea2e4463c5b8536a8b)

ptersilie commented:
@ltratt back to you good Sir!

ltratt commented:
Please squash.

ptersilie commented:
Sorry, another cfmt before I can squash: [4160934](https://github.com/ykjit/yk/pull/578/commits/4160934153311668fe55f1ec40b479678037a5d2)

ltratt commented:
Please squash.

ptersilie commented:
Squashed. But we can't merge until https://github.com/ykjit/ykllvm/pull/48 is in. But it's getting late now. So maybe we should continue tomorrow.

vext01 commented:
Is this ready to go now?

ptersilie commented:
Yes. Let's merge!

vext01 commented:
bors r+

ptersilie commented:
Sorry. This needs another squash.

ltratt commented:
Please squash (and rebase against master).

ptersilie commented:
Squashed and rebased.

vext01 commented:
bors r+
